src/RDom.h

```diff
@@ -243,6 +243,88 @@ class RDom {
     /** Single-dimensional reduction domains can be also be used as Exprs directly. */
     EXPORT operator Expr() const;
 
+    /** Add a predicate to the RDom. An RDom may have multiple
+     * predicates associated with it. An update definition that uses
+     * an RDom only iterates over the subset points in the domain for
+     * which all of its predicates are true. The predicate expression
+     * obeys the same rules as the expressions used on the
+     * right-hand-side of the corresponding update definition. It may
+     * refer to the RDom's variables and free variables in the Func's
+     * update definition. It may include calls to other Funcs, or make
+     * recursive calls to the same Func. This permits iteration over
+     * non-rectangular domains, or domains with sizes that vary with
+     * some free variable, or domains with shapes determined by some
+     * other Func.
+     *
+     * Note that once RDom is used in the update definition of some
+     * Func, no new predicates can be added to the RDom.
+     *
+     * Consider a simple example:
+     \code
+     RDom r(0, 20, 0, 20);
+     r.where(r.x < r.y);
+     r.where(r.x == 10);
+     r.where(r.y > 13);
+     f(r.x, r.y) += 1;
+     \endcode
+     * This is equivalent to:
+     \code
+     for (int r.y = 0; r.y < 20; r.y++) {
+       if (r.y > 13) {
+         for (int r.x = 0; r.x < 20; r.x++) {
+           if (r.x == 10) {
+             if (r.x < r.y) {
+               f[r.x, r.y] += 1;
+             }
+           }
+         }
+       }
+     }
+     \endcode
+     *
+     * Where possible Halide restricts the range of the containing for
+     * loops to avoid the cases where the predicate is false so that
+     * the if statement can be removed entirely. The case above would
+     * be further simplified into:
+     *
+     \code
+     for (int r.y = 14; r.y < 20; r.y++) {
+       f[r.x, r.y] += 1;
+     }
+     \endcode
+     *
+     * In general, the predicates that we can simplify away by
+     * restricting loop ranges are inequalities that compare an inner
+     * Var or RVar to some expression in outer Vars or RVars.
+     *
+     * You can also pack multiple conditions into one predicate like so:
+     *
+     \code
+     RDom r(0, 20, 0, 20);
+     r.where((r.x < r.y) && (r.x == 10) && (r.y > 13));
+     f(r.x, r.y) += 1;
+     \endcode
+     *
+     * This produces the following loop nest:
+     *
+     \code
+     for (int r.y = 0; r.y < 20; r.y++) {
+       for (int r.x = 0; r.x < 20; r.x++) {
+         if ((r.x < r.y) && (r.x == 10) && (r.y > 13)) {
+           f[r.x, r.y] += 1;
+         }
+       }
+     }
+     \endcode
+     *
+     * Using separate RDom::where calls is preferable however, since
```

Comment:
fixed. now we're tracking AND of expr instead of vector of exprs. We split the big expr into ANDs during lowering and inject IfThenElse for each vector element.